WcaOnRails/app/controllers/regional_organizations_controller.rb

```diff
@@ -10,6 +10,14 @@ def admin
 
   def index
     @acknowledged_regional_organizations = RegionalOrganization.currently_acknowledged.order(country: :asc)
+    @img_paths = []
+    @acknowledged_regional_organizations.each do |ro|
```

Comment:
Ruby has strong support for functional operations. You can do
```ruby
@img_paths = @acknowledged_regional_organizations.map do |ro|
  # ...stuff...
end
```

Reply:
That being said, in terms of a 100% clean implementation I think it would be reasonable to consider adjusting the `serializable_hash` for the `RegionalOrganization` model to include its own `thumbnail_url` (or an empty string) as part of the serialized data.

Passing around a second array that has the exact same indices as the original data collection seems a bit overkill...

Reply:
I tried rewriting `serializable_hash`, but `polymorphic_url` is not an available function in models. Perhaps I can find a way to make it available or not even need it?

Reply:
Hrm... Controllers inherit from a base class which in turn internally includes modules (think of it as packages) that provide `polymorphic_url`. You're right that in a model you'd probably have to import that specifc module yourself.

After some more digging, `ro.logo.variant(....).processed.url` seems like it could do the trick without even needing to invoke polymorphic_url at all. See if that works, otherwise just leave as-is.

Reply:
Kudos for finding the proper `url_helper` methods! However, I have a bad feeling about using this in production.

Essentially, right now you found a way to make it work under the exact assumption that files are uploaded to the local storage (which is what happens in local testing). In real prod use, these assumptions will break because we actually store RO stuff on an Amazon S3 bucket. Especially the `only_path: true` part will mess things up there.

The reason why most URL methods do not work in models is because they rely on the _convention_ (i.e. unspoken silent assumption) that they're being called in controllers. Certain module imports and config values are only set in controllers, not in model classes. However, I feel like there has to be some way of "properly" including an asset URL in a model's serial hash.

Summoning @jonatanklosko @viroulep for help :pray:

Reply:
There's https://guides.rubyonrails.org/active_storage_overview.html#serving-files which recommends using `url_for`. Yet again, that's a helper only available in views and controllers, not in models :/

Reply:
From what I understand `rails_representstion_url` generates an internal URL, but that endpoint redirects the resource, which could be S3 or anywhere else. Based on some search the current approach looks reasonable, see [this discussion](https://github.com/rails/rails/issues/32500#issuecomment-380004250). I think we can use `rails_representation_path` instead of `only_path`.

As for using route helpers, we rarely want to do it in models, but calling `Rails.application.routes.url_helpers` is alright.

Reply:
Thank you for the input! `rails_representation_url` is working, but a host url needs to be set. For dev environment, this is simply done with `Rails.application.routes.default_url_options[:host] = 'localhost:3000'`. I wonder what I should put in place of `localhost:3000` for production, `s3.us-west-2.amazonaws.com/www.worldcubeassociation.org`?

Reply:
@notrichardpeng what is the generated path? It seems to me that this URL should always be internal (so in production it would be `https://worldcubeassociation.org/*` and that redirects to S3). If so, using `rails_representation_path` should do the job. I may totally be wrong on this though :)

Reply:
@jonatanklosko This is the link of the image if it's what you mean by generated path: `http://localhost:3000/rails/active_storage/disk/.../random.png`. I went to our production site and copied the image address and it is `https://www.worldcubeassociation.org/rails/active_storage/representations/redirect/.../XXX.png`. Not sure if the small difference is negligible

Reply:
And if you go to `https://www.worldcubeassociation.org/rails/active_storage/representations/redirect/.../XXX.png` it redirects to the image on S3?

Reply:
yes! `https://s3.us-west-2.amazonaws.com/www.worldcubeassociation.org/variants/...`

Reply:
Cool, so we can just use `rails_representation_path`, this way it's going to be `/rails/active_storage/representations/redirect/.../XXX.png` and should work both on localhost and in prod without additional configuration.